### ISSUE TITLE
Naming conventions for packages and modules around Membrane Core

### ISSUE THREAD

**Issue by bblaszkow06** [membraneframework/membrane_core#236]
**Current solution:**
- membrane_aac_format package, with Membrane.AAC.Description (former caps) and Membrane.AAC.Packet (that is going to be used instead of Membrane.Buffer and can be added later, as a separate task). The package is aimed to be lightweight.
- membrane_aac package with Membrane.AAC.Filler (in general, the place for all the membrane elements, that do not rely on any big/native/problematic library)
- membrane_aac_fdk package with Membrane.AAC.FDK.{Decoder, Encoder} (and everything AAC-specific that relies on FDK)

Name changes for old packages are listed in [this post](https://github.com/membraneframework/membrane-core/issues/236#issuecomment-601187727)

---

Currently, the packages are named as follows:
* `membrane-element-XXX` provide element(s) handling some format, e.g. `membrane-element-fdk-aac` that providing AAC decoder & encoder based on Fraunhofer FDK AAC libraries.
  * The modules are named `Membrane.Element.X`, e.g `Membrane.Element.FDK.AAC.Decoder`
* `membrane-caps-{audio,video}-X` that provide caps module for the specified format, e.g. `Membrane.Caps.Audio.AAC`

### `element` in package name
Packages usually provide more than one element, this name may be misleading

### `Element` in module name
Elements added by packages are under `Membrane.Element` namespace while `Membrane.Element` is for the end user something completly different from, say, `Membrane.Element.FDK.AAC.Decoder`

### Element modules should not be too long
Although usually the modules are used only once, so not a key requirement

### Element modules should not share same suffix
This breaks aliasing, but may be solved be rename on alias, so not a big issue

### Namespace conflicts
Modules from different packages should rather live in separate namespaces (with different prefix)
In particular, we probably don't want to put caps/format under the same namespace as elements.

### `Caps` term
This GStreamer-derived term (short for capabilities) is now used for struct describing stream metadata. We need to introduce a clear disctintion:
1. Each pad supports some set of formats. This SET of formats should be called 'caps'
2. Each format is described by a struct that should be called format

Also, like in [Audio.Raw caps](https://github.com/membraneframework/membrane-caps-audio-raw/blob/master/lib/membrane_caps_audio_raw/format.ex), we sometimes include some helpers besides simple format description and it might be surprising that it's in `caps` module

### `Audio`, `Video`
Sometimes it's obvious and this part is not needed. There won't be `Video.AAC` format as one of the A's stands for Audio. Raw formats could become `AudioRaw` and `VideoRaw`

### The convention should allow third-party elements and format definitions
We assume that `Membrane.` prefix is exclusive to packages provided by us (i.e. that are under `membraneframework` organization). The convention should allow third-party developers to follow it without `polluting` our namespace

## Ideas

Packages:
`membrane-element-fdk-aac` -> `membrane-plugin-fdk-aac`
`membrane-caps-audio-aac` -> `membrane-format-aac`
or
`membrane-element-fdk-aac` -> `membrane-aac-fdk-plugin`
`membrane-caps-audio-aac` -> `membrane-aac-format`

Modules:
`Membrane.Element.FDK.AAC.Decoder` -> 
`Membrane.Plugin.FdkAac.{Decoder, Encoder}` or
`Membrane.FdkAacPlugin.{Decoder, Encoder}`

`Membrane.Caps.Audio.AAC` -> `Membrane.Format.AAC` (doesn't solve 3rd-party problem)

**Comment by mat-hek** [comment]
More ideas:
`Membrane.Element.AAC` -> `Membrane.AAC`
`Membrane.Element.FDK.AAC` -> `Membrane.AAC.FDK`
`Membrane.Caps.Audio.AAC` -> `Membrane.AAC.Format`

The main problem with the naming of third-party packages is that it's our `Membrane` prefix what determines that there is some membrane stuff inside, while we don't wish third-party packages to use that prefix. Thus, it's not clear by the name alone that `SomePackage.AAC.Format` is actually a membrane format. The similar problem concerns elements, bins and pipelines. It would be really useful to have a global, concise way of marking modules that can be used only within Membrane. Plugs usually use `.Plug` suffix to solve that, and we could go that way, but this would require us to add suffixes everywhere. I'm not sure we want that, especially when we already have the `Membrane.` prefix.

**Comment by mkaput** [comment]
What about grouping packages into, say, *themes*/*topics* where usually a *theme*/*topic* would correspond to particular codec/codec family/protocol, for example `aac`, `rtp`, `hls`.

Then our packages could be named as follow: `membrane-aac-fdk`, `membrane-rtp`

While namespaces (note there is a bit of inconsistency here but I think this might be a very minor tradeoff):

- `Membrane.*` for core stuff
- `MembraneAAC.*` with `MembraneAAC.FDKDecoder`, `MembraneAAC.AACFormat`
- `MembraneRTP.*` ...

One more thing I have observed is that format packages are incredibly small and have no dependencies. I wonder if we could collect all standard formats into one uber-package `membrane-formats`.

Pros:
- This would reduce amount of dependencies for end-users to have by a half I think
- For us this will reduce amount of repositories to maintain
- I believe having excess format modules is almost (if entirely) costless
- This modules will rarely change in time, unless some minor improvements like docs or revisions.
Cons:
- ?

**Comment by mat-hek** [comment]
> MembraneAAC.* with MembraneAAC.FDKDecoder, MembraneAAC.AACFormat

Decoder and Format cannot reside in the same package, as it would require the user to have all its (possibly native) dependencies to do anything with the given format.

Concerning uber `membrane-formats`, it's an interesting idea I also thought about. However, it's not true that it would reduce the number of dependencies -  all caps/formats required by any element are already specified in its package, so they don't have to be explicitly specified by the end-user. When it comes to versioning, maintaining older versions of some formats would be impossible and that may be required at some point. That would also make it impossible to have a single compatibility source. Don't think that having fewer repositories to maintain (which is the only advantage left) is really worth that, especially when we have Sebex coming in.

**Comment by mkaput** [comment]
> Decoder and Format cannot reside in the same package, as it would require the user to have all its (possibly native) dependencies to do anything with the given format.

Yup, hence I said this will be a bit inconsistent: multiple hex packages would share same namespace. But I do not see this as a huge tradeoff as these packages would be complementary and fully controlled by same entity anyway (excluding malicious behaviours of others of course which are possible anyway).

> However, it's not true that it would reduce the number of dependencies - all caps/formats required by any element are already specified in its package, so they don't have to be explicitly specified by the end-user.

Yes and no. It will reduce amount of work for mix and hex.

> When it comes to versioning, maintaining older versions of some formats would be impossible and that may be required at some point.

The thing is that I wonder if it would ever happen. We are not going to include unstable (in development) formats there but rather standardized ones for long time. The only case I can imagine then is that we will release a bug which would have to be fixed anyway by breaking the continuity of the world if the change would be severe.

**Comment by mat-hek** [comment]
> multiple hex packages would share same namespace. But I do not see this as a huge tradeoff

agreed

> It will reduce amount of work for mix and hex.

Not really beneficial...

> The thing is that I wonder if it would ever happen.

I wonder too :) but formats somehow reflect standards that can be really inconsistently shaped/versioned (such as MPEG) and when we add another format conforming to such standard we may realize we need to rearrange existing formats for it all to make any sense. It has already happened and I can imagine it will do again.

> We are not going to include unstable (in development) formats there but rather standardized ones for long time

While in development, we can end up with multiple branches in this uber package, and projects using some weird merges of different variations of those branches until there's a release.

**Comment by mat-hek** [comment]
So we had another discussion with @bblaszkow06 and agreed on the following:
- membrane-aac-format package, with Membrane.AAC.Description (former caps) and Membrane.AAC.Packet (that is going to be used instead of Membrane.Buffer and can be added later, as a separate task). The package is aimed to be lightweight.
- membrane-aac package with Membrane.AAC.Filler (in general, the place for all the membrane elements, that do not rely on any big/native/problematic library)
- membrane-aac-fdk package with Membrane.AAC.FDK.{Decoder, Encoder}  (and everything AAC-specific that relies on FDK)

As adding a plug-like suffix is problematic, we accept it's enough for third-party libraries to use the names `.Description`, `.Packet`, `.Pipeline`, etc for membrane stuff. We should also place some instructions for contributing to the readme/guide, where we should suggest sticking to our packaging convention.

This is all to be discussed during the meeting tomorrow.

**Comment by mat-hek** [comment]
During today's meeting, we agreed on the version from my post above and concluded that:
- The name `Description` could be more appropriate, so we'll try to come up with a better one
- We considered removing the third dot from module names, so they could be like `Membrane.AAC.FDKEncoder`, but agreed that leaving it (`Membrane.AAC.FDK.Encoder`) is more flexible
- We also discussed and rejected the following options: `MembraneAAC.*`, `Membrane.AACFDK.*`, `Mbrn.AAC.*`, `membrane-aac-plugin`
- The new convention will be applied to the new packages from now, but won't release them till all the old packages are renamed

The names of already published packages will change as follows:

membrane_bin_rtp -> membrane_rtp with Membrane.RTP.{Receiver, Session, JitterBuffer, Parser, ...} and Membrane.RTCP.*
membrane_caps_audio_flac -> membrane_flac_format
membrane_caps_audio_mpeg -> membrane_mp3_format
membrane_caps_audio_raw -> membrane_rawaudio_format
membrane_caps_rtp -> membrane_rtp_format
membrane_caps_video_h264 -> membrane_h264_format
membrane_caps_video_raw -> membrane_rawvideo_format
membrane_element_fake -> membrane_fake
membrane_element_fdk_aac -> membrane_aac_fdk
membrane_element_ffmpeg_h264 -> membrane_h264_ffmpeg
membrane_element_ffmpeg_swresample -> membrane_rawaudio_ffmpeg with Membrane.RawAudio.FFmpeg.Resampler
membrane_element_file -> membrane_file
membrane_element_flac_parser -> membrane_flac with Membrane.FLAC.Parser
membrane_element_gcloud_speech_to_text -> membrane_gclound_speech_to_text (or membrane_gcloud_stt?)
membrane_element_hackney -> membrane_hackney
membrane_element_ibm_speech_to_text -> membrane_ibm_speech_to_text (membrane_ibm_stt ?)
membrane_element_lame -> membrane_mp3_lame
membrane_element_mad -> membrane_mp3_mad
membrane_element_mpegaudioparse -> membrane_mp3 with Membrane.MP3.Parser
membrane_element_portaudio -> membrane_portaudio
membrane_element_rawvideo_parser -> membrane_rawvideo with Membrane.RawVideo.Parser
membrane_element_rtp -> membrane_rtp
membrane_element_rtp_h264 -> membrane_rtp_h264
membrane_element_rtp_jitter_buffer -> membrane_rtp with Membrane.RTP.JitterBuffer
membrane_element_rtp_mpeguadio -> membrane_rtp_mp3
membrane_element_sdl -> membrane_sdl
membrane_element_tee -> membrane_tee
membrane_element_udp -> membrane_udp
membrane_protocol_sdp -> membrane_sdp

**Comment by mkaput** [comment]
> membrane_element_hackney -> membrane_hackney

If this element is not explicitly exposing Hackney APIs/Internals (which I believe should not), then maybe it could be called `membrane_http`. That would open potential possibility to replace underlying HTTP library if necessary.

---

Regarding `Description` I think I would go with `Format` anyway because of three reasons:

1. it would match the names of the packages (`-format`)
2. It seems to be more widely used as a name for entities describing, well, formats. I spot that in DirectShow, FFmpeg (AVInputFormat, AVOutputFormat), Audacity (File/SampleFormat), Ableton Live docs
3. It's shorter

**Comment by mat-hek** [comment]
I agree about Hackney, updated that post, does the author agree too @bblaszkow06 ?

Regarding description, the first point is rather a disadvantage, because if we had Membrane.AAC.Format in the package membrane-aac-format, then it would be a bit counter-intuitive if there was something not namespaced under Membrane.AAC.Format there, such as Membrane.AAC.Packet. However, I could live with that anyway.

**Comment by mat-hek** [comment]
I have another idea regarding the format/description - what if we just had Membrane.AAC? I mean Membrane.Caps.AAC -> Membrane.AAC? That would introduce much confusion as Membrane.AAC would reside in membrane-aac-format instead of membrane-aac, so we could go the @bblaszkow06 way with `-plugin` suffix:
- membrane-aac with Membrane.AAC and Membrane.AAC.Packet
- membrane-aac-plugin with Membrane.AAC.Parser etc
- membrane-aac-fdk-plugin with Membrane.AAC.FDK.Decoder etc

We could also use `-kit` / `-tool{kit, set, box}` / `-suite` / ? instead of `-plugin`

**Comment by bblaszkow06** [comment]
1. It had `_hackney` in the name exactly to be able to switch it. It was a replacement for element `httpoison`. Naming it `membrane_http` will limit us to providing only one http client. So it should rather be `membrane_http_hackney` unless we're 100% sure we won't need to maintain both.
Also, having `membrane_aac_fdk` we might want to keep it consistent and go with `membrane_http_hackney`
---
1. Note that it contains more than just this format description. If the name matched the package exactly, the presence of other things could be a bit surprising.
2. Actually, you provided an argument against that name:
    * in FFmpeg `fomrat` is for FILE formats. AV{In,Out}putFormat classes are basically interfaces for muxing and demuxing. It's hard to map FFmpeg to our abstractions, but this description would be a combination of FFmpeg's metadata from AVXFormat or AVCodec
and AVFrames/AVPackets
    * Audacity - that's raw audio sample format. That means we will have `RawAudio.Format` struct containing field `format` (then we should probably rename it to longer `sample_format`)
 
Plus, Apple's Core Media framework uses [CMFormatDescription](https://developer.apple.com/documentation/coremedia/cmformatdescription-u8g) which seems pretty similar to what we have here

Considering the above I'm having doubts about using "format" for a package name. And I would prefer MediaDescription over Description, but the followers of the WeShouldUseAsLittleLettersAsHumanlyPosible Church would probably kill me :wink:

**Comment by mat-hek** [comment]
Concerning hackney, note that if we had previously named HTTPoison just HTTP, we could have replaced HTTPoison with Hackney seamlessly.  BTW make sure you've seen my previous post, as we posted at exactly the same time :P

**Comment by bblaszkow06** [comment]
1. That would be still inconsistent with other packages as we always use lib name and I can imagine a scenario where 2 different implementations of HTTP client element would be maintained (e.g. if we added libcurl, then as all packages with native deps it would require suffix and we would probably like to keep pure Elixir version as well).
2. > Concerning 'format', IMHO we shouldn't reject it just because FFmpeg uses it somewhere

That's not what happened here. FFmpeg was used as an argument for naming it `Format` and that argument was invalid.

**Comment by mat-hek** [comment]
Ok, changed to `membrane-http-hackney'. We will stick to [what we agreed on](https://github.com/membraneframework/membrane-core/issues/236#issuecomment-601187727) and this will be the final version unless we manage to agree on some changes before we need to release membrane-rtp.

**Comment by mat-hek** [comment]
I have another proposition (regarding what we have not discussed yet): we could change repositories names to use underscores instead of dashes, like `membrane-core` -> `membrane_core`, `membrane-aac-format` -> `membrane_aac_format`. Despite they would be slightly more inconvenient to write (need for pressing shift to write an underscore), they would match names of packages on Hex. That would also make them conform to the `mix new` generator convention.

**Comment by bblaszkow06** [comment]
Working at RTP plugin I came to conclusion that using just `Membrane.RTP` instead of `Membrane.RTP. Description` would look much better (just like @mat-hek proposed in https://github.com/membraneframework/membrane-core/issues/236#issuecomment-601384906).
However, what I didn't like about that idea was to use bare `membrane_rtp` for package name of format description - I'd stick to `membrane_X_format` and `membrane_X_plugin`.

Regarding the `_plugin` alternatives:
* `suite` - I would rule it out as it may be confused with test suite
* `kit`/`toolkit` - it wouldn't make sense in some cases, e.g. `membrane_tee_toolkit` seems off while `membrane_tee_plugin` as something that extends framework capabilities by providing tee(s) is still valid.

The last point gives me another idea - `plugin` could be replaced by `extension` but for me they are both equally fine.
